rita_exit/src/database/mod.rs

```diff
@@ -517,3 +526,88 @@ pub fn enforce_exit_clients(client_data: &mut RitaExitData) -> Result<(), Box<Ri
     client_data.set_debt_actions(new_debt_actions);
     Ok(())
 }
+
+/// Removes client ip assignments for clients that have not checked in for 1 hour
+pub fn teardown_inactive_clients(client_data: &mut RitaExitData) {
+    let inactive_list = client_data.get_inactive_list();
+    // check for snat mode
+    let ipv4_mode = client_data.get_ipv4_nat_mode();
+    if let ExitIpv4RoutingSettings::SNAT { .. } = ipv4_mode {
+    } else {
+        return;
+    }
+    let ext_nic = &settings::get_rita_exit().network.external_nic.unwrap();
+    let ext_assignments = client_data.get_external_ip_assignments();
+    let int_assignments = client_data.get_internal_ip_assignments();
+    // for each client in inactive, find their wg key from clientandipmap, then
+    // remove their iptables rules if their last seen was over threshold
+    for (client, last_seen) in inactive_list.iter() {
+        if last_seen.elapsed().as_secs() > INACTIVE_CLIENT_THRESHOLD {
+            try_teardown_client_snat(
+                ext_assignments.clone(),
+                int_assignments.clone(),
+                *client,
+                ext_nic,
+            );
+        }
+    }
+    let all_clients = client_data.get_all_registered_clients();
+    // then, list any clients that have not had a wg handshake recently
+    let wg_clients = get_wg_exit_clients_offline("wg_exit").unwrap();
+    for wg_key in wg_clients.iter() {
+        let client = all_clients
+            .iter()
+            .find(|c| *wg_key == c.wg_public_key)
+            .expect("Client not found in database");
+        try_teardown_client_snat(
+            ext_assignments.clone(),
+            int_assignments.clone(),
+            *client,
+            ext_nic,
+        );
+    }
+}
+
+fn try_teardown_client_snat(
+    ext_assignments: HashMap<Ipv4Addr, HashSet<Identity>>,
+    int_assignments: HashMap<Ipv4Addr, Identity>,
+    client: Identity,
+    ext_nic: &str,
+) {
+    // get this client's assigned external ip
+    let external_ip: Vec<Ipv4Addr> = ext_assignments
+        .iter()
+        .filter(|(_k, v)| v.contains(&client))
+        .map(|(&k, _v)| k)
+        .collect();
+    let internal_ip: Vec<Ipv4Addr> = int_assignments
+        .iter()
+        .filter(|(_k, v)| **v == client)
+        .map(|(&k, _v)| k)
+        .collect();
+    match (external_ip.len().cmp(&1), internal_ip.len().cmp(&1)) {
+        (std::cmp::Ordering::Greater, _) | (_, std::cmp::Ordering::Greater) => {
```

Comment:
this function is a little dense with all the functional operations, maybe break it out into a few little utility functions so that it's easier to see how exactly it works. It's not clear to me what happens in the error cases.

Also if the teardown fails we shouldn't release the ip back into the pool to be used, potentially resulting in a not working client connecting, better to properly return an error and keep a hold of the ip until the teardown succeeds.